### ISSUE TITLE
Data parallel inference

### ISSUE THREAD

**Issue by kevinhu** [vllm-project/vllm#1237]
Is there a recommended way to run data parallel inference (i.e. a copy of the model on each GPU)? It's possible by hacking CUDA_VISIBLE_DEVICES, but I was wondering if there's a cleaner method.
```python
def worker(worker_idx):
    os.environ["CUDA_VISIBLE_DEVICES"] = str(worker_idx)
    prompts = [
        "Hello, my name is",
        "The president of the United States is",
        "The capital of France is",
        "The future of AI is",
    ]
    sampling_params = SamplingParams(temperature=0.8, top_p=0.95)
    llm = LLM(model="facebook/opt-125m")
    outputs = llm.generate(prompts, sampling_params)


if __name__ == "__main__":
    
    with multiprocessing.Pool(4) as pool:
        pool.map(worker, range(4))
```

**Comment by viktor-ferenczi** [comment]
This approach should end up in a more scalable (maybe also cleaner) architecture:

Run a vLLM API server for each GPU, serving on different ports. Then use those API endpoints to schedule generations from a centralized process on the vLLM backends. If you want to do it from Python, then try the `vllm-client` package, it supports async, making your logic simpler. 

It also allows restarting your "control" process during development (or on upgrades) without having to reload the model into the vLLM backends. 

It has a slight overhead due to API access, but that would amortize due to the parallel execution of generations.

Make sure to keep the vLLM backend processes alive. Restart them if they crash or if they repeatedly fail client requests.

**Comment by viktor-ferenczi** [comment]
**Feature request:** Allow for data-parallel execution on multiple (sets of) GPUs with the same model, served from the same API, so no external scheduler is required.

**Comment by brucechin** [comment]
> This approach should end up in a more scalable (maybe also cleaner) architecture:
> 
> Run a vLLM API server for each GPU, serving on different ports. Then use those API endpoints to schedule generations on the vLLM backends from a centralized process. If you want to do it from Python, then try the [vllm-client](https://github.com/viktor-ferenczi/vllm-client/) package (also available to install with pip), it supports async, making your logic simpler.
> 
> It also allows restarting your "control" process during development (or on upgrades) without having to reload the model into the vLLM backends.
> 
> It has a slight overhead due to API access, but that would amortize due to the parallel execution of generations.
> 
> Make sure to keep the vLLM backend processes alive. Restart them if they crash or if they repeatedly fail client requests.

Hi @viktor-ferenczi , @LiuXiaoxuanPKU assigned this issue to me. After reading your comments, I think to add the data-parallel inference on multiple GPUs with the same model, I can implement it according to your above suggestion?
1. A centralized scheduler process to start multiple local vLLM API servers, each for each GPU with different ports.
2. Implement some scheduling policies and restart API server when it detects any failure.
3. If possible, we can support multi-server and each server with multi-GPU scenario to further improve the scalability.

Please let me know if I miss anything here. I would like to add this feature support in my free time.

**Comment by viktor-ferenczi** [comment]
I'm not 100% sure that this functionality belongs to the vLLM engine project itself, because it is only a layer on top of it. Maybe using some existing external tool/framework to verify service health and configure it to restart the vLLM instances if required would be enough. All it needs is running a short generation as a health check one a minute (for example), so broken/frozen processes can be identified and restarted automatically.

You have the freedom to go ahead and make a PR for the solution your described.

**Comment by SunLemuria** [comment]
I think fastchat supports this feature: [fastchat scalability](https://github.com/lm-sys/FastChat/blob/main/README.md#optional-advanced-features-scalability)
<img width="1057" alt="image" src="https://github.com/vllm-project/vllm/assets/37478528/8b4e0041-33fa-4dc8-b6e7-cd8d7861436f">

**Comment by anisingh1** [comment]
Hi @brucechin, Are you working on implementing this request or has this been deferred?

**Comment by AjayP13** [comment]
This is possible to do on our `[DataDreamer](https://datadreamer.dev/docs/latest/)` package which can load vLLM in parallel (different models on different GPUs). It does this by always instantiating vLLM in a background process and communicating with it. See `ParallelLLM` in the package for wrapping multiple `VLLM` objects under a single LLM object.

**Comment by darrenglow** [comment]
> > This approach should end up in a more scalable (maybe also cleaner) architecture:
> > Run a vLLM API server for each GPU, serving on different ports. Then use those API endpoints to schedule generations on the vLLM backends from a centralized process. If you want to do it from Python, then try the [vllm-client](https://github.com/viktor-ferenczi/vllm-client/) package (also available to install with pip), it supports async, making your logic simpler.
> > It also allows restarting your "control" process during development (or on upgrades) without having to reload the model into the vLLM backends.
> > It has a slight overhead due to API access, but that would amortize due to the parallel execution of generations.
> > Make sure to keep the vLLM backend processes alive. Restart them if they crash or if they repeatedly fail client requests.
> 
> Hi @viktor-ferenczi , @LiuXiaoxuanPKU assigned this issue to me to offload some work. After reading your comments, I think to add the data-parallel inference on multiple GPUs with the same model, I can implement it according to your above suggestion?
> 
> 1. A centralized scheduler process to start multiple local vLLM API servers, each for each GPU with different ports.
> 2. Implement some scheduling policies and restart API server when it detects any failure.
> 3. If possible, we can support multi-server and each server with multi-GPU scenario to further improve the scalability.
> 
> I plan to add a new class : DataParallelScheduler which can start multiple vLLM API servers, manage them, and schedule incoming requests with the same `generate` interface. In the `vllm/entrypoints/api_server.py`, add an option for data-parallel inference, if enabled, instead of starting `engine = AsyncLLMEngine.from_engine_args(engine_args)`, we can initialize a `DataParallelScheduler` instance for serving the request.
> 
> I will ensure that my change will not affect the old execution flow when the data-parallel inference option is disabled. I will also add tests to check the robustness of the scheduler I am going to add.
> 
> Please let me know if I miss anything here. I would like to add this feature support in my free time.

Hi, @brucechin , how is this work going? I am fascinated about this idea.

**Comment by AmoghM** [comment]
+1 for this feature.

**Comment by zemerov** [comment]
+1 for this feature.

**Comment by WanBenLe** [comment]
+1 for this feature, datadreamer seems couldn't Improve inference speed(with model copy)) of a single model on multiple GPUs

**Comment by ifromeast** [comment]
+1 for this feature. @WoosukKwon

**Comment by GritLs** [comment]
+1  for this feature

**Comment by NewXiaoMeng** [comment]
need this feature too.

**Comment by kota-iizuka** [comment]
There is an example of using data parallel in https://github.com/vllm-project/vllm/blob/main/examples/offline_inference_distributed.py (changing `num_instances` properly controls the CUDA_VISIBLE_DEVICES environment variable as well).

On the other hand, since the above example is batch inference, I think there is still a need for a method of online inference (with proper load balancing) and a simple method for parallel inference of multiple models. (It is probably possible to achieve this with a single script in the examples/ directory, but it is important to make it easy to use.)

**Comment by shizhediao** [comment]
+1 for this feature.

**Comment by zhaochenyang20** [comment]
https://github.com/zhaochenyang20/ModelServer

Could you please check this? This is my locally written one.

**Comment by youkaichao** [comment]
I'm going to close this issue, as vllm does not plan to support it.

users should seek third party support (which should be pretty easy to set up), e.g.:

https://docs.litellm.ai/docs/simple_proxy#load-balancing---multiple-instances-of-1-model 

or the solutions mentioned in the above discussions.

**Comment by vadimkantorov** [comment]
@youkaichao otherwise, should the `multiprocessing.Pool.map` method in the OP be good?

**Comment by vadimkantorov** [comment]
Does the `torchrun` example have the same issue? I just wish to do offline inference of a large set of prompts in the most efficient way

**Comment by vadimkantorov** [comment]
I'm using sth like the example in the OP, but getting in the end the following errors. How can I make it so vllm engines (in multiprocessing.Process instances) shutdown gracefully without these nasty messages?

```
[rank7]:[W528 00:08:15.330160171 TCPStore.cpp:125] [c10d] recvValue failed on SocketImpl(fd=190, addr=[localhost]:60524, remote=[localhost]:44431): failed to recv, got 0 bytes
Exception raised from recvBytes at /pytorch/torch/csrc/distributed/c10d/Utils.hpp:678 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x98 (0x7f8c600f45e8 in /mnt/fs/venv_cu126_py312/lib/python3.12/site-packages/torch/lib/libc10.so)
frame #1: <unknown function> + 0x5ba8afe (0x7f8c464deafe in /mnt/fs/venv_cu126_py312/lib/python3.12/site-packages/torch/lib/libtorch_cpu.so)
frame #2: <unknown function> + 0x5baae40 (0x7f8c464e0e40 in /mnt/fs/venv_cu126_py312/lib/python3.12/site-packages/torch/lib/libtorch_cpu.so)
frame #3: <unknown function> + 0x5bab74a (0x7f8c464e174a in /mnt/fs/venv_cu126_py312/lib/python3.12/site-packages/torch/lib/libtorch_cpu.so)
frame #4: c10d::TCPStore::check(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) + 0x2a9 (0x7f8c464db1a9 in /mnt/fs/venv_cu126_py312/lib/python3.12/site-packages/torch/lib/libtorch_cpu.so)
frame #5: c10d::ProcessGroupNCCL::heartbeatMonitor() + 0x379 (0x7f8c036d49a9 in /mnt/fs/venv_cu126_py312/lib/python3.12/site-packages/torch/lib/libtorch_cuda.so)
frame #6: <unknown function> + 0xdc253 (0x7f8d06715253 in /lib/x86_64-linux-gnu/libstdc++.so.6)
frame #7: <unknown function> + 0x94ac3 (0x7f8d0979fac3 in /lib/x86_64-linux-gnu/libc.so.6)
frame #8: <unknown function> + 0x126850 (0x7f8d09831850 in /lib/x86_64-linux-gnu/libc.so.6)

[rank7]:[W528 00:08:15.334364022 ProcessGroupNCCL.cpp:1659] [PG ID 0 PG GUID 0 Rank 7] Failed to check the "should dump" flag on TCPStore, (maybe TCPStore server has shut down too early), with error: failed to recv, got 0 bytes
[rank5]:[W528 00:08:15.642241559 TCPStore.cpp:125] [c10d] recvValue failed on SocketImpl(fd=184, addr=[localhost]:60508, remote=[localhost]:44431): failed to recv, got 0 bytes
Exception raised from recvBytes at /pytorch/torch/csrc/distributed/c10d/Utils.hpp:678 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x98 (0x7f8c600f45e8 in /mnt/fs/venv_cu126_py312/lib/python3.12/site-packages/torch/lib/libc10.so)
frame #1: <unknown function> + 0x5ba8afe (0x7f8c464deafe in /mnt/fs/venv_cu126_py312/lib/python3.12/site-packages/torch/lib/libtorch_cpu.so)
frame #2: <unknown function> + 0x5baae40 (0x7f8c464e0e40 in /mnt/fs/venv_cu126_py312/lib/python3.12/site-packages/torch/lib/libtorch_cpu.so)
frame #3: <unknown function> + 0x5bab74a (0x7f8c464e174a in /mnt/fs/venv_cu126_py312/lib/python3.12/site-packages/torch/lib/libtorch_cpu.so)
frame #4: c10d::TCPStore::check(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) + 0x2a9 (0x7f8c464db1a9 in /mnt/fs/venv_cu126_py312/lib/python3.12/site-packages/torch/lib/libtorch_cpu.so)
frame #5: c10d::ProcessGroupNCCL::heartbeatMonitor() + 0x379 (0x7f8c036d49a9 in /mnt/fs/venv_cu126_py312/lib/python3.12/site-packages/torch/lib/libtorch_cuda.so)
frame #6: <unknown function> + 0xdc253 (0x7f8d06715253 in /lib/x86_64-linux-gnu/libstdc++.so.6)
frame #7: <unknown function> + 0x94ac3 (0x7f8d0979fac3 in /lib/x86_64-linux-gnu/libc.so.6)
frame #8: <unknown function> + 0x126850 (0x7f8d09831850 in /lib/x86_64-linux-gnu/libc.so.6)

[rank3]:[W528 00:08:15.642258697 TCPStore.cpp:125] [c10d] recvValue failed on SocketImpl(fd=178, addr=[localhost]:60492, remote=[localhost]:44431): failed to recv, got 0 bytes
Exception raised from recvBytes at /pytorch/torch/csrc/distributed/c10d/Utils.hpp:678 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x98 (0x7f8c600f45e8 in /mnt/fs/venv_cu126_py312/lib/python3.12/site-packages/torch/lib/libc10.so)
frame #1: <unknown function> + 0x5ba8afe (0x7f8c464deafe in /mnt/fs/venv_cu126_py312/lib/python3.12/site-packages/torch/lib/libtorch_cpu.so)
frame #2: <unknown function> + 0x5baae40 (0x7f8c464e0e40 in /mnt/fs/venv_cu126_py312/lib/python3.12/site-packages/torch/lib/libtorch_cpu.so)
frame #3: <unknown function> + 0x5bab74a (0x7f8c464e174a in /mnt/fs/venv_cu126_py312/lib/python3.12/site-packages/torch/lib/libtorch_cpu.so)
frame #4: c10d::TCPStore::check(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) + 0x2a9 (0x7f8c464db1a9 in /mnt/fs/venv_cu126_py312/lib/python3.12/site-packages/torch/lib/libtorch_cpu.so)
frame #5: c10d::ProcessGroupNCCL::heartbeatMonitor() + 0x379 (0x7f8c036d49a9 in /mnt/fs/venv_cu126_py312/lib/python3.12/site-packages/torch/lib/libtorch_cuda.so)
frame #6: <unknown function> + 0xdc253 (0x7f8d06715253 in /lib/x86_64-linux-gnu/libstdc++.so.6)
frame #7: <unknown function> + 0x94ac3 (0x7f8d0979fac3 in /lib/x86_64-linux-gnu/libc.so.6)
frame #8: <unknown function> + 0x126850 (0x7f8d09831850 in /lib/x86_64-linux-gnu/libc.so.6)

[rank5]:[W528 00:08:15.646333013 ProcessGroupNCCL.cpp:1659] [PG ID 0 PG GUID 0 Rank 5] Failed to check the "should dump" flag on TCPStore, (maybe TCPStore server has shut down too early), with error: failed to recv, got 0 bytes
[rank3]:[W528 00:08:15.646362708 ProcessGroupNCCL.cpp:1659] [PG ID 0 PG GUID 0 Rank 3] Failed to check the "should dump" flag on TCPStore, (maybe TCPStore server has shut down too early), with error: failed to recv, got 0 bytes
```

**Comment by teja-metrum-ai** [comment]
```
vllm-serving-1  | INFO 06-05 07:11:00 [__init__.py:239] Automatically detected platform rocm.
vllm-serving-1  | INFO 06-05 07:11:01 [api_server.py:1034] vLLM API server version 0.8.3.dev349+gb8498bc4a
vllm-serving-1  | INFO 06-05 07:11:01 [api_server.py:1035] args: Namespace(subparser='serve', model_tag='meta-llama/Llama-3.3-70B-Instruct', config='', host=None, port=8000, uvicorn_log_level='info', disable_uvicorn_access_log=False, allow_credentials=False, allowed_origins=['*'], allowed_methods=['*'], allowed_headers=['*'], api_key=None, lora_modules=None, prompt_adapters=None, chat_template=None, chat_template_content_format='auto', response_role='assistant', ssl_keyfile=None, ssl_certfile=None, ssl_ca_certs=None, enable_ssl_refresh=False, ssl_cert_reqs=0, root_path=None, middleware=[], return_tokens_as_token_ids=False, disable_frontend_multiprocessing=False, enable_request_id_headers=False, enable_auto_tool_choice=False, tool_call_parser=None, tool_parser_plugin='', model='meta-llama/Llama-3.3-70B-Instruct', task='auto', tokenizer=None, hf_config_path=None, skip_tokenizer_init=False, revision=None, code_revision=None, tokenizer_revision=None, tokenizer_mode='auto', trust_remote_code=True, allowed_local_media_path=None, download_dir=None, load_format='auto', config_format=<ConfigFormat.AUTO: 'auto'>, dtype='auto', kv_cache_dtype='auto', max_model_len=8192, guided_decoding_backend='xgrammar', logits_processor_pattern=None, model_impl='auto', distributed_executor_backend=None, pipeline_parallel_size=1, tensor_parallel_size=1, data_parallel_size=2, enable_expert_parallel=False, max_parallel_loading_workers=None, ray_workers_use_nsight=False, block_size=None, enable_prefix_caching=False, prefix_caching_hash_algo='builtin', disable_sliding_window=False, use_v2_block_manager=True, num_lookahead_slots=0, seed=None, swap_space=4, cpu_offload_gb=0, gpu_memory_utilization=0.9, num_gpu_blocks_override=None, max_num_batched_tokens=320000, max_num_partial_prefills=1, max_long_partial_prefills=1, long_prefill_token_threshold=0, max_num_seqs=None, max_logprobs=20, disable_log_stats=False, quantization=None, rope_scaling=None, rope_theta=None, hf_token=None, hf_overrides=None, enforce_eager=False, max_seq_len_to_capture=8192, disable_custom_all_reduce=False, tokenizer_pool_size=0, tokenizer_pool_type='ray', tokenizer_pool_extra_config=None, limit_mm_per_prompt=None, mm_processor_kwargs=None, disable_mm_preprocessor_cache=False, enable_lora=False, enable_lora_bias=False, max_loras=1, max_lora_rank=16, lora_extra_vocab_size=256, lora_dtype='auto', long_lora_scaling_factors=None, max_cpu_loras=None, fully_sharded_loras=False, enable_prompt_adapter=False, max_prompt_adapters=1, max_prompt_adapter_token=0, device='auto', num_scheduler_steps=1, use_tqdm_on_load=True, multi_step_stream_outputs=True, scheduler_delay_factor=0.0, enable_chunked_prefill=None, speculative_config=None, model_loader_extra_config=None, ignore_patterns=[], preemption_mode=None, served_model_name=None, qlora_adapter_name_or_path=None, show_hidden_metrics_for_version=None, otlp_traces_endpoint=None, collect_detailed_traces=None, disable_async_output_proc=False, scheduling_policy='fcfs', scheduler_cls='vllm.core.scheduler.Scheduler', override_neuron_config=None, override_pooler_config=None, compilation_config=None, kv_transfer_config=None, worker_cls='auto', worker_extension_cls='', generation_config='auto', override_generation_config=None, enable_sleep_mode=False, calculate_kv_scales=False, additional_config=None, enable_reasoning=False, reasoning_parser=None, disable_cascade_attn=False, disable_log_requests=True, max_log_len=None, disable_fastapi_docs=False, enable_prompt_tokens_details=False, enable_server_load_tracking=False, dispatch_function=<function ServeSubcommand.cmd at 0x7f076eac0040>)
vllm-serving-1  | INFO 06-05 07:11:19 [config.py:604] This model supports multiple tasks: {'generate', 'classify', 'reward', 'embed', 'score'}. Defaulting to 'generate'.
vllm-serving-1  | INFO 06-05 07:11:19 [arg_utils.py:1735] rocm is experimental on VLLM_USE_V1=1. Falling back to V0 Engine.
vllm-serving-1  | INFO 06-05 07:11:24 [api_server.py:246] Started engine process with PID 156
vllm-serving-1  | INFO 06-05 07:11:27 [__init__.py:239] Automatically detected platform rocm.
vllm-serving-1  | INFO 06-05 07:11:28 [llm_engine.py:242] Initializing a V0 LLM engine (v0.8.3.dev349+gb8498bc4a) with config: model='meta-llama/Llama-3.3-70B-Instruct', speculative_config=None, tokenizer='meta-llama/Llama-3.3-70B-Instruct', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, override_neuron_config=None, tokenizer_revision=None, trust_remote_code=True, dtype=torch.bfloat16, max_seq_len=8192, download_dir=None, load_format=LoadFormat.AUTO, tensor_parallel_size=1, pipeline_parallel_size=1, disable_custom_all_reduce=False, quantization=None, enforce_eager=False, kv_cache_dtype=auto,  device_config=cuda, decoding_config=DecodingConfig(guided_decoding_backend='xgrammar', reasoning_backend=None), observability_config=ObservabilityConfig(show_hidden_metrics=False, otlp_traces_endpoint=None, collect_model_forward_time=False, collect_model_execute_time=False), seed=None, served_model_name=meta-llama/Llama-3.3-70B-Instruct, num_scheduler_steps=1, multi_step_stream_outputs=True, enable_prefix_caching=False, chunked_prefill_enabled=False, use_async_output_proc=True, disable_mm_preprocessor_cache=False, mm_processor_kwargs=None, pooler_config=None, compilation_config={"splitting_ops":[],"compile_sizes":[],"cudagraph_capture_sizes":[256,248,240,232,224,216,208,200,192,184,176,168,160,152,144,136,128,120,112,104,96,88,80,72,64,56,48,40,32,24,16,8,4,2,1],"max_capture_size":256}, use_cached_outputs=True, 
vllm-serving-1  | INFO 06-05 07:11:35 [rocm.py:181] None is not supported in AMD GPUs.
vllm-serving-1  | INFO 06-05 07:11:35 [rocm.py:182] Using ROCmFlashAttention backend.
vllm-serving-1  | INFO 06-05 07:11:35 [parallel_state.py:836] Adjusting world_size=2 rank=0 distributed_init_method=tcp://127.0.0.1:59905 for DP
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448] Timed out after 601 seconds waiting for clients. 1/2 clients joined.
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448] Traceback (most recent call last):
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]   File "/usr/local/lib/python3.12/dist-packages/vllm/engine/multiprocessing/engine.py", line 436, in run_mp_engine
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]     engine = MQLLMEngine.from_vllm_config(
vllm-serving-1  | Process SpawnProcess-1:
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]   File "/usr/local/lib/python3.12/dist-packages/vllm/engine/multiprocessing/engine.py", line 128, in from_vllm_config
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]     return cls(
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]            ^^^^
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]   File "/usr/local/lib/python3.12/dist-packages/vllm/engine/multiprocessing/engine.py", line 82, in __init__
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]     self.engine = LLMEngine(*args, **kwargs)
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]                   ^^^^^^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]   File "/usr/local/lib/python3.12/dist-packages/vllm/engine/llm_engine.py", line 281, in __init__
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]     self.model_executor = executor_class(vllm_config=vllm_config, )
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]   File "/usr/local/lib/python3.12/dist-packages/vllm/executor/executor_base.py", line 52, in __init__
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]     self._init_executor()
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]   File "/usr/local/lib/python3.12/dist-packages/vllm/executor/uniproc_executor.py", line 46, in _init_executor
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]     self.collective_rpc("init_device")
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]   File "/usr/local/lib/python3.12/dist-packages/vllm/executor/uniproc_executor.py", line 56, in collective_rpc
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]     answer = run_method(self.driver_worker, method, args, kwargs)
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]   File "/usr/local/lib/python3.12/dist-packages/vllm/utils.py", line 2533, in run_method
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]     return func(*args, **kwargs)
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]            ^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]   File "/usr/local/lib/python3.12/dist-packages/vllm/worker/worker_base.py", line 604, in init_device
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]     self.worker.init_device()  # type: ignore
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]     ^^^^^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]   File "/usr/local/lib/python3.12/dist-packages/vllm/worker/worker.py", line 194, in init_device
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]     init_worker_distributed_environment(self.vllm_config, self.rank,
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]   File "/usr/local/lib/python3.12/dist-packages/vllm/worker/worker.py", line 533, in init_worker_distributed_environment
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]     init_distributed_environment(parallel_config.world_size, rank,
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]   File "/usr/local/lib/python3.12/dist-packages/vllm/distributed/parallel_state.py", line 844, in init_distributed_environment
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]     torch.distributed.init_process_group(
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]   File "/usr/local/lib/python3.12/dist-packages/torch/distributed/c10d_logger.py", line 81, in wrapper
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]     return func(*args, **kwargs)
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]            ^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]   File "/usr/local/lib/python3.12/dist-packages/torch/distributed/c10d_logger.py", line 95, in wrapper
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]     func_return = func(*args, **kwargs)
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]                   ^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]   File "/usr/local/lib/python3.12/dist-packages/torch/distributed/distributed_c10d.py", line 1710, in init_process_group
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]     store, rank, world_size = next(rendezvous_iterator)
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]                               ^^^^^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]   File "/usr/local/lib/python3.12/dist-packages/torch/distributed/rendezvous.py", line 230, in _tcp_rendezvous_handler
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]     store = _create_c10d_store(
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]             ^^^^^^^^^^^^^^^^^^^
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]   File "/usr/local/lib/python3.12/dist-packages/torch/distributed/rendezvous.py", line 198, in _create_c10d_store
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]     return TCPStore(
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448]            ^^^^^^^^^
vllm-serving-1  | ERROR 06-05 07:21:36 [engine.py:448] torch.distributed.DistStoreError: Timed out after 601 seconds waiting for clients. 1/2 clients joined.
vllm-serving-1  | Traceback (most recent call last):
vllm-serving-1  |   File "/usr/lib/python3.12/multiprocessing/process.py", line 314, in _bootstrap
vllm-serving-1  |     self.run()
vllm-serving-1  |   File "/usr/lib/python3.12/multiprocessing/process.py", line 108, in run
vllm-serving-1  |     self._target(*self._args, **self._kwargs)
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/vllm/engine/multiprocessing/engine.py", line 450, in run_mp_engine
vllm-serving-1  |     raise e
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/vllm/engine/multiprocessing/engine.py", line 436, in run_mp_engine
vllm-serving-1  |     engine = MQLLMEngine.from_vllm_config(
vllm-serving-1  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/vllm/engine/multiprocessing/engine.py", line 128, in from_vllm_config
vllm-serving-1  |     return cls(
vllm-serving-1  |            ^^^^
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/vllm/engine/multiprocessing/engine.py", line 82, in __init__
vllm-serving-1  |     self.engine = LLMEngine(*args, **kwargs)
vllm-serving-1  |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/vllm/engine/llm_engine.py", line 281, in __init__
vllm-serving-1  |     self.model_executor = executor_class(vllm_config=vllm_config, )
vllm-serving-1  |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/vllm/executor/executor_base.py", line 52, in __init__
vllm-serving-1  |     self._init_executor()
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/vllm/executor/uniproc_executor.py", line 46, in _init_executor
vllm-serving-1  |     self.collective_rpc("init_device")
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/vllm/executor/uniproc_executor.py", line 56, in collective_rpc
vllm-serving-1  |     answer = run_method(self.driver_worker, method, args, kwargs)
vllm-serving-1  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/vllm/utils.py", line 2533, in run_method
vllm-serving-1  |     return func(*args, **kwargs)
vllm-serving-1  |            ^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/vllm/worker/worker_base.py", line 604, in init_device
vllm-serving-1  |     self.worker.init_device()  # type: ignore
vllm-serving-1  |     ^^^^^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/vllm/worker/worker.py", line 194, in init_device
vllm-serving-1  |     init_worker_distributed_environment(self.vllm_config, self.rank,
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/vllm/worker/worker.py", line 533, in init_worker_distributed_environment
vllm-serving-1  |     init_distributed_environment(parallel_config.world_size, rank,
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/vllm/distributed/parallel_state.py", line 844, in init_distributed_environment
vllm-serving-1  |     torch.distributed.init_process_group(
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/torch/distributed/c10d_logger.py", line 81, in wrapper
vllm-serving-1  |     return func(*args, **kwargs)
vllm-serving-1  |            ^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/torch/distributed/c10d_logger.py", line 95, in wrapper
vllm-serving-1  |     func_return = func(*args, **kwargs)
vllm-serving-1  |                   ^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/torch/distributed/distributed_c10d.py", line 1710, in init_process_group
vllm-serving-1  |     store, rank, world_size = next(rendezvous_iterator)
vllm-serving-1  |                               ^^^^^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/torch/distributed/rendezvous.py", line 230, in _tcp_rendezvous_handler
vllm-serving-1  |     store = _create_c10d_store(
vllm-serving-1  |             ^^^^^^^^^^^^^^^^^^^
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/torch/distributed/rendezvous.py", line 198, in _create_c10d_store
vllm-serving-1  |     return TCPStore(
vllm-serving-1  |            ^^^^^^^^^
vllm-serving-1  | torch.distributed.DistStoreError: Timed out after 601 seconds waiting for clients. 1/2 clients joined.
vllm-serving-1  | Traceback (most recent call last):
vllm-serving-1  |   File "/usr/local/bin/vllm", line 8, in <module>
vllm-serving-1  |     sys.exit(main())
vllm-serving-1  |              ^^^^^^
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/vllm/entrypoints/cli/main.py", line 51, in main
vllm-serving-1  |     args.dispatch_function(args)
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/vllm/entrypoints/cli/serve.py", line 27, in cmd
vllm-serving-1  |     uvloop.run(run_server(args))
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/uvloop/__init__.py", line 109, in run
vllm-serving-1  |     return __asyncio.run(
vllm-serving-1  |            ^^^^^^^^^^^^^^
vllm-serving-1  |   File "/usr/lib/python3.12/asyncio/runners.py", line 195, in run
vllm-serving-1  |     return runner.run(main)
vllm-serving-1  |            ^^^^^^^^^^^^^^^^
vllm-serving-1  |   File "/usr/lib/python3.12/asyncio/runners.py", line 118, in run
vllm-serving-1  |     return self._loop.run_until_complete(task)
vllm-serving-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  |   File "uvloop/loop.pyx", line 1518, in uvloop.loop.Loop.run_until_complete
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/uvloop/__init__.py", line 61, in wrapper
vllm-serving-1  |     return await main
vllm-serving-1  |            ^^^^^^^^^^
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/vllm/entrypoints/openai/api_server.py", line 1069, in run_server
vllm-serving-1  |     async with build_async_engine_client(args) as engine_client:
vllm-serving-1  |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  |   File "/usr/lib/python3.12/contextlib.py", line 210, in __aenter__
vllm-serving-1  |     return await anext(self.gen)
vllm-serving-1  |            ^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/vllm/entrypoints/openai/api_server.py", line 146, in build_async_engine_client
vllm-serving-1  |     async with build_async_engine_client_from_engine_args(
vllm-serving-1  |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  |   File "/usr/lib/python3.12/contextlib.py", line 210, in __aenter__
vllm-serving-1  |     return await anext(self.gen)
vllm-serving-1  |            ^^^^^^^^^^^^^^^^^^^^^
vllm-serving-1  |   File "/usr/local/lib/python3.12/dist-packages/vllm/entrypoints/openai/api_server.py", line 269, in build_async_engine_client_from_engine_args
vllm-serving-1  |     raise RuntimeError(
vllm-serving-1  | RuntimeError: Engine process failed to start. See stack trace for the root cause.
vllm-serving-1 exited with code 1
```
it is exiting even for data-parallel 1 , any suggestion for running the data parallel right way
